pystiche/papers/ulyanov_et_al_2016/data.py

```diff
@@ -0,0 +1,293 @@
+from typing import Optional, Sized
+from urllib.parse import urljoin
+import torch
+from torch.utils.data import Dataset, Sampler, DataLoader
+from pystiche.data import (
+    DownloadableImage,
+    DownloadableImageCollection,
+    CreativeCommonsLicense,
+    PublicDomainLicense,
+    ImageFolderDataset,
+    FiniteCycleBatchSampler,
+)
+from pystiche.image import extract_num_channels, CaffePreprocessing
+from pystiche.image.transforms import Transform, Resize, ComposedTransform
+from pystiche.image.transforms.functional import grayscale_to_fakegrayscale
+
+
+__all__ = [
+    "ulyanov_et_al_2016_content_transform",
+    "ulyanov_et_al_2016_style_transform",
+    "ulyanov_et_al_2016_images",
+    "ulyanov_et_al_2016_dataset",
+    "ulyanov_et_al_2016_batch_sampler",
+    "ulyanov_et_al_2016_image_loader",
+]
+
+
+def ulyanov_et_al_2016_content_transform(
+    edge_size: int = 256, impl_mode: str = "paper"
+) -> ComposedTransform:
+    class OptionalGrayscaleToFakegrayscale(Transform):
+        def forward(self, input_image: torch.Tensor) -> torch.Tensor:
+            is_grayscale = extract_num_channels(input_image) == 1
+            if is_grayscale:
+                return grayscale_to_fakegrayscale(input_image)
+            else:
+                return input_image
+
+    transforms = []
+    if impl_mode == "texturenetsv1":
+        transforms.append(Resize((edge_size, edge_size), interpolation_mode="bilinear"))
+    elif impl_mode == "paper":
+        transforms.append(
+            Resize((edge_size, edge_size), interpolation_mode="bilinear")
+        )  # FIXME: paper?
+    elif impl_mode == "master":
+        # FIXME: RandomCrop here
```

Comment:
Please address this `FIXME`.

Reply:
The author of the paper has used RandomCrop (master branch) or has not given any information on pre-processing in his paper. Should therefore a scaling take place before the RandomCrop in this implementation?

Reply:
If no information is given in the paper we should always assume the simplest case. In this case that means that no preprocesing is perfomerd at all.

Reply:
This will not be possible here. Due to the structure of the generator network and the associated downscaling and upscaling of the content image, the image should have a size which is a multiple of 2. For example 256x256 or 512x512. For this reason, the generated images in the paper also have the size of 256x256.

Reply:
If the images used in the paper are `256x256` why do we need cropping at all?

Reply:
The authors use the images from the _ImageNet  ILSVRC2012_ and the resolution of these images varies. For this reason they must have preprocessed the images, but there is no information about this step.

Reply:
Ah, OK. In that case I suggest we use `CenterCrop`, since this most of the time captures the motif of the image.